### PR TITLE
MULTIPLAYER: Multiplayer Servers for HE Games.

### PR DESCRIPTION
**Pull request by LittleToonCat**
This is a companion PR to a larger PR in the main source repository implementing Network Play for Humongous Entertainment titles:
scummvm/scummvm/pull/4725

This consists of three servers, a session server, a web server, and a lobby server, each with their own purposes.

- Session Server:
This server (located in main.py) is the main event.  It allows users to host and join an online game session from over the Internet, and assists the host with hole-punching.  It can also relay game traffic if NAT-traversal fails.  It uses Redis for session hosting and keeps track of relaying users.

- Web Server:
This aims to show all the available and joinable sessions, it looks basic and messy at the moment as I don't know a lot of HTML/CSS, but it'll probably get all situated eventually.  It gathers session information from Redis which gets added by the session server. 

- Lobby Server:
This is the original and main server code used for [Backyard Sports Online](https://backyardsports.online) located [here.](https://github.com/Backyard-Sports-Online/server), it can communicate to a website API to verify user logins and change user stats.  It has been modified and improved upon, mainly with the addition of TLS hosting.  It once again uses Redis to keep track of user logins and area management.

The README.md file talk all about how to run them yourself, and the default servers coded in the source PR is up and running at the time of writing, so have at it; feedback appreciated.

I'll "un-draft" this PR when the source PR is ready to merge.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,63 @@
...
-Currently supports multiplayer games for Moonbase Commander
+This project contains code for hosting online multiplayer lobbies for compatable Humongous Entertainment games.
 
+## Getting Started
+### Installing
+Both the session and web servers requires Redis to be installed.  This is needed for both servers to share session data to each other.
+
+Clone this repo and checkout to the multiplayer branch:
+```
+git clone https://github.com/LittleToonCat/scummvm-sites.git
```
Comment (sev-) [inline README.md]: This needs to b updated

File: lobby/database/Redis.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
+"use strict";
```
Comment (sev-) [inline lobby/database/Redis.js:22]: I think it would be nice to have the copyright headers to all files

**Comment by rvanlaar** [comment]
Hello,

Thank you for taking the work to program all of this.
I've quickly taken a look at the python part.

I saw some typing, which is nice. 

I have two areas of improvement. 
Could you run a formatter such as black over the code? All indentation is done with tabs, python recommends using spaces.

Please use `if __name__ == "__main__" to wrap all top level python code.
<https://docs.python.org/3/library/__main__.html>

File: lobby/global/Stats.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    // TODO: Football
+    "football": (resultsFields, isHome, opponentId) => {
+        return {"not_yet_supported": 1}
+    }
+};
+
+module.exports = {
+    ResultsMappers: ResultsMappers,
+};
```
Comment (sev-) [inline lobby/global/Stats.js]: Missing end of line :)

File: net_defines.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Copied from engines/scumm/he/moonbase/net_defines.h
+
+# These are used for relaying.
+PN_PRIORITY_HIGH = 			0x00000001
+PN_SENDTYPE_INDIVIDUAL =	1
+PN_SENDTYPE_GROUP =			2
+PN_SENDTYPE_HOST =			3
+PN_SENDTYPE_ALL =			4
```
Comment (sev-) [inline net_defines.py]: no endline

File: main.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+	redis.rpush(f"{game}:sessions", session_id)
+
+	# Store to address to session hash
+	redis.hset(f"{game}:sessionByAddress", str(address), session_id)
+
+	logging.debug(f"{address}: NEW SESSION: \"{name}\"")
+	return session_id
+
+def relay_data(data, sent_peer):
+	from_user = data.get("from")
```
Comment (sev-) [inline main.py]: I wonder how bulletproof our server is. Is there a way to make sure it can cope with malformed input?

File: web/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM tiangolo/uvicorn-gunicorn-fastapi:python3.9
```
Comment (sev-) [inline web/Dockerfile]: For safety reasons, I would consider forking this to our repo and pulling from there or switching to some non-private docker image source.

File: web/config.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	"football2002": "Backyard Football 2002",
+	"moonbase": "Moonbase Commander"
+}
+
+# Version variants to accept for a specific game.
+# If none exist but game exist in the GAMES list,
+# that means there's only one game version.
+VERSIONS = {
+	"moonbase": ["1.0", "1.1", "Demo"]
+}
```
Comment (sev-) [inline web/config.py]: ...end line

File: web/static/style.css [hunk 8]
```diff
@@ -59,15 +64,33 @@ a.link > img {
...
+    background-color: #cccccc;
+  }
+
 @media only screen and (max-width: 640px) {
   a.link > b {
     font-size: 10pt;
-}
-}
+  }
+}
```
Comment (sev-) [inline web/static/style.css]: ...end line

File: web/templates/game.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                        <dd>- Version: {{session["version"]}}</dd>
+                    {% endif -%}
+                {% endfor -%}
+                </dl>
+                {% endif -%}
+            {% endif -%}
+            </div>
+        </div>
+    </body>
+</html>
```
Comment (sev-) [inline web/templates/game.html]: ...end line

File: web/templates/index.html [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+                <a href="/{{game}}" class="link">
+                    <img src="{{'/static/icons/{}.png'.format(game)}}" />
+                    <b>{{names[game]}}</b>
+                </a>
+                {% endfor -%}
+                </div>
+            </div>
+        </div>
+    </body>
+</html>
```
Comment (sev-) [inline web/templates/index.html]: ...end line

File: web/templates/index.html [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+<html>
+    <head>
+        <title>ScummVM Multiplayer</title>
+        <link rel="stylesheet" type="text/css" href="/static/style.css"/>
+        <link rel="icon" type="image/x-icon" href="https://scummvm.org/favicon.ico">
+    </head>
+    <body>
+        <div class="container">
+            <div class="header">
+                <img src="https://scummvm.org/images/scummvm_logo.png"/>
```
Comment (sev-) [inline web/templates/index.html]: Do we need this cross-site image? Why not host it right here?

**Review by sev- — COMMENTED** [review]
I did some cosmetic review.

**Comment by Thunderforge** [comment]
The setup instructions in `README.md` seem incomplete. I ran these three without issues:

```
python3 -m venv .env
source .env/bin/activate
python3 -m pip install -r requirements.txt
```

Then tried to start it up with
```
python3 main.py
```

Which results in

```
redis.exceptions.ConnectionError: Error 61 connecting to 127.0.0.1:6379. Connection refused.
```

So it seems a database is supposed to be set up, but isn't.

File: lobby/database/Redis.js (hunk 2 above)
Comment (LittleToonCat) [reply lobby/database/Redis.js:22]: Including the Python files?

File: main.py (hunk 5 above)
Comment (LittleToonCat) [reply main.py]: I've been testing it thoroughly throughout development, and I feel like it should be stable enough.  Nonetheless, I've added sanity checks for game data to make sure they are correct before relaying them out: 98a37f4411a165ef139b12f6a7fa1f4a81fb7e98

**Pull request merged**